tests/compute/test_kernel.py

```diff
@@ -0,0 +1,235 @@
+import dgl
+import dgl.function as fn
+import networkx as nx
+import backend as F
+from itertools import product
+
+def udf_src_x_edge(broadcast):
+    def fn(edges):
+        n = edges.src['n']
+        e = edges.data['e']
+        if broadcast == 'src':
+            n = F.unsqueeze(n, 1)
+        elif broadcast == 'edge':
+            e = F.unsqueeze(e, 1)
+        return {'m' : n * e}
+    return fn
+
+def udf_copy_src(edges):
+    return {'m' : edges.src['n']}
+
+def udf_copy_edge(edges):
+    return {'m' : edges.data['e']}
+
+def udf_sum(nodes):
+    return {'r1' : nodes.mailbox['m'].sum(1)}
+
+def udf_max(nodes):
+    return {'r1' : F.max(nodes.mailbox['m'], 1)}
+
+D1 = 5
+D2 = 10
+D3 = 4
+builtin = {'sum': fn.sum, 'max': fn.max}
+udf_reduce = {'sum': udf_sum, 'max': udf_max}
+fill_value = {'sum': 0, 'max': float("-inf")}
+
+def generate_graph(broadcast='none'):
+    """Create graph with src, edge, dst feature. broadcast can be 'src',
+    'edge', 'dst', 'none'
+    """
+    g = dgl.DGLGraph(nx.erdos_renyi_graph(100, 0.5))
+    nv = g.number_of_nodes()
+    ne = g.number_of_edges()
+    if broadcast == 'edge':
+        n = F.randn((nv, D1, D2, D3))
+        e = F.randn((ne, D2, 1))
+        f = F.randn((nv, D1, D2, D3))
+    elif broadcast == 'src':
+        n = F.randn((nv, D2, 1))
+        e = F.randn((ne, D1, D2, D3))
+        f = F.randn((nv, D1, D2, D3))
+    elif broadcast == 'dst':
+        n = F.randn((nv, D1, D2, D3))
+        e = F.randn((ne, D1, D2, D3))
+        f = F.randn((nv, D2, 1))
+    else:
+        n = F.randn((nv, D1, D2, D3))
+        e = F.randn((ne, D1, D2, D3))
+        f = F.randn((nv, D1, D2, D3))
+
+    g.ndata['n'] = F.attach_grad(n)
+    g.ndata['f'] = F.attach_grad(f)
+    g.edata['e'] = F.attach_grad(e)
+    return g
+
+def test_src_op_edge_reduce():
+    def _test(red, broadcast="none"):
+        g = generate_graph(broadcast)
+
+        with F.record_grad():
+            g.update_all(fn.src_mul_edge(src='n', edge='e', out='m'),
+                         builtin[red](msg='m', out='r1'))
+            r1 = g.ndata['r1']
+            F.backward(r1.sum())
+            n_grad1 = F.grad(g.ndata['n'])
+            e_grad1 = F.grad(g.edata['e'])
+
+        # reset grad
+        F.attach_grad(g.ndata['n'])
+        F.attach_grad(g.edata['e'])
+
+        with F.record_grad():
+            g.update_all(udf_src_x_edge(broadcast), udf_reduce[red])
+            r2 = g.ndata['r1']
+            F.backward(r2.sum())
+            n_grad2 = F.grad(g.ndata['n'])
+            e_grad2 = F.grad(g.edata['e'])
+
+        assert F.allclose(r1, r2)
+        assert(F.allclose(n_grad1, n_grad2))
+        assert(F.allclose(e_grad1, e_grad2))
+
+    _test('sum')
+    _test('sum', 'src')
+    _test('sum', 'edge')
+    _test('max')
+    _test('max', 'src')
+    _test('max', 'edge')
```

Comment:
why do we need this test? `test_all_binary_builtins` doesn't cover this case?

Reply:
This one is an old test. The 'edge', 'src' is specifying that broadcasting should be tested on edge or src feature. We can have all comprehensive tests (including broadcasting dimension) for all builtin combinations in test_all_binary_builtins. It's just going to take longer for testing to finish. But yeah it's worth it. Will do.